spack environment (2): expose a subset of packages to a user

citibeth opened the pull request:
This is a continuation of #4585, moving the branch into the main Spack repo.  My latest commits have fixed bugs that became apparent in trying to use Spack Environments.

scheibelp commented on lib/spack/spack/spec.py:
```diff
@@ -1560,6 +1567,19 @@ def from_node_dict(node):
             spec.external_path = None
             spec.external_module = None
 
+        if 'concrete' in node:
```
@citibeth if this is addressing https://github.com/spack/spack/pull/4585#issuecomment-382903792, then this looks like it may be a more general bug and the fix ought to go in its own PR. Would you mind doing that?

citibeth commented:
Yes it is.  But until now, nobody was pickling / unpickling CONCRETIZED
specs to JSON.  So the bug wasn't uncovered until we started doing Spack
Environments.  For that reason, I think it's best to keep the bug fix with
this PR.



On Thu, Apr 19, 2018 at 8:27 PM, scheibelp <notifications@github.com> wrote:

> *@scheibelp* commented on this pull request.
> ------------------------------
>
> In lib/spack/spack/spec.py:
>
> > @@ -1560,6 +1567,19 @@ def from_node_dict(node):
>              spec.external_path = None
>              spec.external_module = None
>
> +        if 'concrete' in node:
>
> @citibeth <https://github.com/citibeth> if this is addressing #4585
> (comment)
> <https://github.com/spack/spack/pull/4585#issuecomment-382903792>, then
> this looks like it may be a more general bug and the fix ought to go in its
> own PR. Would you mind doing that?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spack/spack/pull/7843#pullrequestreview-113819065>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cd-4-zcBaGT_B0braXhB0sAZXP8OHks5tqStigaJpZM4Tcp0T>
> .
>

citibeth commented:
More TODO

- [ ] `spack env uninstall` that works like `spack uninstall`

citibeth commented:
@scheibelp @tgamblin I've made a bunch of improvements to this PR:

* Many `spack env` commands have analogs in the non-env Spack world, and the two should generally work the same way.  For example, `spack env install` should work the same as `spack install`.  Same for `spack env uninstall` vs. `spack uninstall`, `spack env list` vs. `spack spec`, `spack env loads` vs `spack module loads`,   etc.  I've refactored things a bit so the `spack env` versions of these commands gain access to all the command line arguments of the main commands on which they are based.  **HOWEVER**... This turns `env.py` into a dependency magnet, an issue maybe we want to address at some point.  (Actually, `env.py` was already a dependency magnet).

* I added `-o` to `spack env install`, which runs the install from the given directory -- defaulting to the environment dir.  This could affect the version of some Spack build logs.  More importantly, it affects the location of the `spack setup` files (#7846), placing them where they belong in the environment by default.

* I changed `spack env list-packages` to `spack env loads`, and harmonized it with `spack module loads`.  I added the `-o` flag, defaulting to writing the module load script in the Environment directory.

* I added `spack env uninstall`.  As with `spack install`, this interfaces with the underlying `spack uninstall` in the logical manner.

* I added `spack env location`, which just prints the location of an environment.  Marginally useful, but at least it's simple.

QUESTIONS
-------------

1. It looks like every time the environment is changed, the entire Spack Environment directory is rewritten.  This could cause problems if I'm generating `spack setup` files or `spack env loads` files  in there,.  Any thoughts on how best to address this issue?

2. `spack env uninstall` doesn't do what I really want.  For example, I might want to: (a) "Uninstall this environment, but only packages in it that nothing else relies upon", or (b) Garbage Collection: "Uninstall all packages not used by a managed environment"  ("managed environment" means environments stored inside spack/var/spack/environments).  This kind of Garbage Collection would finally give us a convenient way to prune our Spack builds.

3. `spack env location` should be extended so that `spack env <env> location <package>` works like `spack location`.  Your thoughts?

citibeth commented:
TODO
------
- [ ] `spack env view`

scheibelp commented:
> spack env view

I think this should go in its own PR

> I changed `spack env list-packages` to `spack env loads`, and harmonized it with `spack module loads`.

citibeth commented:
Fine with me.  I don't use Spack views, and am unlikely to be motivated to
create `spack env view`.  But my sense of symmetry tells me it should exist.

On Mon, Apr 23, 2018 at 1:38 PM, scheibelp <notifications@github.com> wrote:

> spack env view
>
> I think this should go in its own PR
>
> I changed spack env list-packages to spack env loads, and harmonized it
> with spack module loads.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spack/spack/pull/7843#issuecomment-383659939>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cdxRQy6lBHL9JFrKGI8u2GQdG89sBks5trhGjgaJpZM4Tcp0T>
> .
>

scheibelp commented:
@citibeth there are some flake issues

Generally I think syncing environment commands with install in terms of a consistent API is desirable, although I think it's good to consider when it is necessary here vs. possible in a follow-up. It has the potential to introduce new controversies that slow this PR down. For example while I like the effect produced by adding `recurse_dependencies` to `Spec.tree`, it feels like there's a bit of concept drift with the name (`Spec.tree` that is).

scheibelp commented:
> It looks like every time the environment is changed, the entire Spack Environment directory is rewritten. This could cause problems if I'm generating spack setup files or spack env loads files in there,. Any thoughts on how best to address this issue?

I think the easiest way to handle this would be to have a special directory in the environment called `usr`; the environment makes sure to preserve this when updating, either by copying all the files under `usr` or by moving them to the updated environment directory.

citibeth commented:
@scheibelp  OK, I hacked the reading/writing of environments today.  To try to summarize (supposing <env> is the root directory of an environment; say, `spack/var/spack/environments/foo`):

1. Spack currently only writes into `<env>/.env/environment.json`
2. In the future, Spack will also write a repo into `<env>/.env`.  It can go right in that directory if you like, or in a subdirectory `<en>/.env/repo`.
3. `spack env install` symlinks log files into `<env>/logs`.  This is a really convenient place for the user to check build logs.
4. `spack loads` writes to `<env>/loads`.  There is no longer an option to write to a general filename.  (It introduced complexity for a feature I think will rarely or ever be used.  I like the idea of `spack env` commands modifying a single environment directory in a consistent way).

Please give a holler if you don't like any of the UI, filename, etc. choices here.  I should get a chance to really test this out in the next day or two.

Next I will go on to better management of Environment configs.  (I think I want two options: either create 1 or more configs in the environment, or symlink configs from elsewhere).

citibeth commented:
@tgamblin  Flake8 complained on the new `schema/env.py` file.  But since `env.py` follows the same format as other `schema` files, I'm suspecting that Flake8 is not configured as intended.  Any advice?
```
lib/spack/spack/schema/env.py:42: [E203] whitespace before ':'
lib/spack/spack/schema/env.py:43: [E203] whitespace before ':'
lib/spack/spack/schema/env.py:44: [E203] whitespace before ':'
lib/spack/spack/schema/env.py:45: [E203] whitespace before ':'
lib/spack/spack/schema/env.py:45: [E203] whitespace before ':'
```

scheibelp commented:
It's complaining that

`'configs' : {`

should be

`'configs': {`

(for example)

scheibelp commented on lib/spack/spack/cmd/env.py:
```diff
@@ -22,57 +22,765 @@
+def environment_spec(args):
+    environment = read(args.environment)
+    prepare_repository(environment)
+    prepare_config_scope(environment)
```
@citibeth I assumed your intention was to add a command which works like `spack spec` but uses the environment (and in the future optionally the repository) of the environment. I added these commands to produce that effect.

citibeth commented:
@scheibelp When two specs have conflicting versions of the same package, the first one takes precedence over the later ones when running `spack env loads`.   This is backwards of the precedence order for merged configurations.  I would suggest that the two be harmonized; otherwise, you get weird stuff like the `env.yaml` file below.  The order of `config` precedence seems a little more intuitive to me.

Your thoughts?

```
env:
    configs:
        - '{HOME}/spscopes/centos7'            # Lowest precedence
        - '{HOME}/spscopes/gissversions'
        - '{HOME}/spscopes/twoway'
        - '{HOME}/spscopes/develop'            # Highest precedence
    specs:
        modele-tests:                          # Highest precedence
            setup: [modele-tests]
        modele:
            setup: [modele, ibmisc, pism, icebin]
        modele-control:
        modele-utils:
        ncview:
        netcdf:
        nco:                                   # Lowest precedence
 ```

citibeth commented:
@scheibelp @tgamblin This PR has now served me successfully to install my 95-package software stack.  I'm not 100% happy with it and have a lot of ideas to make it easier / better / more convenient, etc.  But what we have now gets the job done and is (usually) not too inconvenient.  So I would suggest we call Spack Environments a Beta feature (meaning the UI could change), do a feature freeze on this PR, and get it merged.

Your thoughts?

scheibelp commented on lib/spack/docs/environments.rst:
```diff
@@ -0,0 +1,305 @@
+
+.. note::
+
+   Currently, only "bare" user specs may be added to an environment,
```
@citibeth this was supposed to be fixed by https://github.com/spack/spack/pull/7843/commits/d537b69e3cb53c06f77e858b76e1fef4a19ee59d. Does it still not work for you?

citibeth replied:
OK, I removed the text.

scheibelp commented:
> Suppose I have package builtin:P. When that package gets copied into the environment's repo, it should maintain the same name builtin:P. It shouldn't suddenly change namespace.

I thought from our discussion that this issue was acceptable since the `--use-env-repo` was off by default: the idea is that normally you can reconcretize as much as you want and the dag hashes are preserved but if you copy the environment directory to another spack instance you can reconcretize with the packages maintained in the environment. Or is this an additional issue I overlooked?

citibeth commented:
I think it's fine for now.  Later, I think we can make it more useful with
enough clever thinking.

On Wed, May 9, 2018 at 10:00 PM, scheibelp <notifications@github.com> wrote:

> Suppose I have package builtin:P. When that package gets copied into the
> environment's repo, it should maintain the same name builtin:P. It
> shouldn't suddenly change namespace.
>
> I thought from our discussion that this issue was acceptable since the
> --use-env-repo was off by default: the idea is that normally you can
> reconcretize as much as you want and the dag hashes are preserved but if
> you copy the environment directory to another spack instance you can
> reconcretize with the packages maintained in the environment. Or is this an
> additional issue I overlooked?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spack/spack/pull/7843#issuecomment-387928422>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cd_LDR91jXxHJUgG26k_RZBygO7nKks5tw584gaJpZM4Tcp0T>
> .
>

citibeth commented:
I just checked over the whole PR.  Looks good to me, let's get this merged please!

scheibelp commented:
@tgamblin @becker33 this is ready for review

citibeth commented:
Superceded by #8231
